mlflow/models/utils.py

```diff
@@ -612,17 +612,27 @@ def _enforce_schema(pf_input: PyFuncInput, input_schema: Schema):
     For tensor-based signatures, we make sure the shape and type of the input matches the shape
     and type specified in model's input schema.
     """
+
+    def _is_scalar(x):
+        return np.isscalar(x) or x is None
+
     if isinstance(pf_input, pd.Series):
         pf_input = pd.DataFrame(pf_input)
     if not input_schema.is_tensor_spec():
         if isinstance(pf_input, (list, np.ndarray, dict, pd.Series, str)):
             try:
-                if isinstance(pf_input, dict) and all(
-                    not isinstance(value, (dict, list)) for value in pf_input.values()
+                if isinstance(pf_input, str):
+                    pf_input = pd.DataFrame([pf_input])
+                elif isinstance(pf_input, dict) and all(
+                    _is_scalar(value) for value in pf_input.values()
+                ):
+                    pf_input = pd.DataFrame([pf_input])
+                elif (
+                    isinstance(pf_input, dict)
+                    and all(isinstance(value, np.ndarray) for value in pf_input.values())
+                    and all(value.size == 1 for value in pf_input.values())
                 ):
-                    pf_input = pd.DataFrame(pf_input, index=[0])
-                elif isinstance(pf_input, str):
-                    pf_input = pd.DataFrame({"inputs": pf_input}, index=[0])
+                    pf_input = pd.DataFrame([pf_input])
```

Comment:
From https://github.com/mlflow/mlflow/blob/1fa3afbf31adfa4d4954ffa2f6c54c6d5468d53d/mlflow/models/utils.py#L612-L614, the logic prior to MLflow 2.3 for this case was to do:

```
pf_input = pd.DataFrame(pf_input)
```

However, it seems we're now doing 

```
pf_input = pd.DataFrame([pf_input])
```

I examined the difference between the two with an example input:

```
pf_input = {"foo": np.array([1]), "bar": np.array([2])}

print(pd.DataFrame([pf_input]))

   foo  bar
0  [1]  [2]

print(pd.DataFrame(pf_input))

   foo  bar
0    1    2
```

These two DataFrames are definitely different.

Doesn't that mean we're still making a backwards incompatible change?

Reply:
this check is completely invalid and not used for any of the new features anyway. Removed it and restricted the list wrapping of a pf_input to only string type (throws in pre-2.3 MLflow) and Dict[str, scalar] (throws in pre-2.3 MLflow). Those are the only two new data structure types that are needed for transformers. Validated in MLflow 2.2 build that the only changes with this are now the addition of supporting pyfunc inputs of string and Dict[str, scalar].